concurrency/DCMetricsUtils.java

```diff
@@ -0,0 +1,481 @@
+
+    /**
+     * Creates a click stream event for the appropriate page and sub page types that are provided as well as any page action which is taken
+     * @param metricId The metric ID string you want to associate with this metric
+     * @param pageType The page type of the page you are recording action on
+     * @param subPageType The sub page type of the page you are recording action on
+     * @param hitType The hit type associated with the metric you want to log
+     * @param pageAction The action taken on the page, if there is no action then just pass in null such as for a page hit
+     * @param refMarker The refMarker associated with the metric.
+     * @return The new click stream event
+     */
+    public DCMetricsUtils createClickStreamMetric(String metricId, String metricName, String pageType, String subPageType, String hitType,
+                                                  String pageAction, String refMarker) {
+        Slog.d(TAG, "Creating ClickStream Metric - "
+            + metricId + ","
+            + pageType + ","
+            + subPageType + ","
+            + hitType + ","
+            + pageAction
+            + " serviceName=" + serviceName
+            + ", methodname=" + methodName
+            + ", customerId=" + dcmConfig.getCustomerId()
+            + ", session-id=" + dcmConfig.getSessionId()
+        );
+        if (!METRIC_INDEX.containsKey(metricId)) {
+            ClickStreamMetricsEvent metricEvent = mFactory.createClickStreamMetricEvent(serviceName, methodName);
+
+            UsageInfo usageInfo = new UsageInfo(pageType, hitType, teamName, siteVariant);
+            if (subPageType != null && !subPageType.isEmpty()) {
+                usageInfo.setSubPageType(subPageType);
+            }
+            if (pageAction != null && !pageAction.isEmpty()) {
+                usageInfo.setPageAction(pageAction);
+            }
+
+            DataPoint refTagDataPoint = new DataPoint(
+                    REFMARKER_DATA_POINT_NAME,
+                    refMarker,
+                    1, DataPointType.CK
+            );
+
+            try {
+                metricEvent.addDataPoint(refTagDataPoint);
+            } catch (MetricsException ex) {
+                Slog.e(TAG, "Unable to set data point in event for ref_marker", ex);
+            }
+
+            metricEvent.setUsageInfo(usageInfo);
+
+            String customerId = dcmConfig.getCustomerId();
+            String sessionId = dcmConfig.getSessionId();
+            if (customerId != null && sessionId != null) {
+                metricEvent.setAnonymous(false);
+                metricEvent.setNonAnonymousCustomerId(customerId);
+                metricEvent.setNonAnonymousSessionId(sessionId);
+            } else {
+                metricEvent.setAnonymous(true);
+            }
+
+            metricName = stripIllegalCharacters(metricName);
+            metricEvent.addCounter(metricName, 1);
+
+            METRIC_INDEX.put(metricId, metricEvent);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **402 and 440**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).